proposals/NNNN-high-level-interface-modal-view-controllers.md

```diff
@@ -0,0 +1,160 @@
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
+
+@property (nonatomic) BOOL shouldPlayTone;
+@property (nonatomic) BOOL progressIndicator;
+
+@end
+``` 
+
+#### SDLAudioInputController
+
+The audio input controller is used to abstract the RPC `PerformAudioPassThru`, `OnAudioPassThru` and `EndAudioPassThru`. The app developer should be able to provide a block which is called for audio data. Furthermore a method called `dismiss` should help to end the audio pass thru.
+
+```objc
+@interface SDLAudioInputController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) SDLSamplingRate samplingRate;
+@property (nonatomic, copy) SDLBitsPerSample bitsPerSample;
+@property (nonatomic, copy) SDLAudioType audioType;
+@property (nonatomic) BOOL muteMediaSource;
+
+@property (nonatomic) (^onAudioData)(NSData *audioData);
+
+- (void)finish; // sends an EndAudioPassThru in order to stop audio input
+
+@end
+```
+
+#### SDLScrollableMessageController
+
+Similar to `Alert` this view controller takes care of the RPC `ScrollableMessage`.
+
+```objc
+@interface SDLScrollableMessageController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
+
+@end
+```
+
+#### SDLSliderController
+
+This controller matches very closely to the view [`UISlider`](https://developer.apple.com/documentation/uikit/uislider) but for consistency to SDL it makes more sense to treat it as a modal controller. In order to work with the RPC `Slider` it is important to provide the selected value back to the caller. As of UIKit this is not done by using a custom completion handler but simply provide the result in a property of the modal view controller. As per mobile API the property `Slider.sliderFooter` is an array used for two purposes. Either it's a footer (single item) or a list of names representing slider values (number of items must match `.numTicks`). The mobile API allows `.numTicks` to be between 2 and 26. The value range for the parameter `.position` can between 1 and 26 which means the range of values is not more than 26. With the high level abstraction the modal controller can manage the specified range allowing different min and max values. Examples are ranges with negative which are mapped internally to the available range. Future versions of the mobile API could increase the range which would be managed by the slider controller.
+
+```objc
+@interface SDLSliderController : SDLModalViewController
+
+@property (nonatomic, copy) NSString *title; // mapped with Slider.sliderHeader
+@property (nonatomic, copy) NSArray<NSString *> *valueLables; // overrides control of SDLModalViewController.message 
+
+// minimumValue and maximumValue map to `Slider.numTicks`
+@property (nonatomic) NSInteger minimumValue;
+@property (nonatomic) NSInteger maximumValue;
+
+// maps to `Slider.position`
+@property (nonatomic) NSInteger value;
```

Comment:
`initialPosition`?

Reply:
different to other properties `.value` should not be read-only. It should cover the request's param `.position` as well as the response's param `.sliderPosition`. Needs documentation that programmatically changing the value during presentation will be ignored/not be applied.

Reply:
How does the VC get notified that it's done? It's through the completion handler on the VC manager? Then the property gets re-inspected? I can see this being confusing for developers, even though it's similar to how `UISlider` works, because it's still significantly different due to differences between UIKit and SDL (the UIKit view stays on screen vs. the SDL is modal only and will disappear, the value changes as the user changes it vs. the value is only updated when the view disappears). Unless there's a special initializer that can provide a callback when the value changes (e.g. on the response)? I'm not sure of the best way to approach it.